homeassistant/components/stream/__init__.py

```diff
@@ -91,17 +91,18 @@
 
 _LOGGER = logging.getLogger(__name__)
 
-STREAM_SOURCE_REDACT_PATTERN = [
-    (re.compile(r"//.*:.*@"), "//****:****@"),
-    (re.compile(r"\?auth=.*"), "?auth=****"),
-]
-
 
-def redact_credentials(data: str) -> str:
+def redact_credentials(url: str) -> str:
     """Redact credentials from string data."""
-    for (pattern, repl) in STREAM_SOURCE_REDACT_PATTERN:
-        data = pattern.sub(repl, data)
-    return data
+    yurl = URL(url)
+    if yurl.user is not None:
+        yurl = yurl.with_user("****")
+    if yurl.password is not None:
+        yurl = yurl.with_password("****")
+    redacted_query_params = {
+        key: "****" for key in {"auth", "user", "password"} & yurl.query.keys()
+    }
```

Comment:
Good place to use the oft-overlooked `dict.fromkeys` function:
```suggestion
    redacted_query_params = dict.fromkeys(("auth", "user", "password", *yurl.query), "****")
```

Reply:
Thanks, I will use dict.with_keys instead of a dict generator. But this suggestion will end up changing the behavior to add the keys even if they are not present in the original query params.